### ISSUE TITLE
Capability declaration / detection on dialects and engines

### ISSUE THREAD

**Issue by arlyon** [tortoise/tortoise-orm#86]
This was mentioned in the GIS discussion but never made it into a seperate issue.

Having a client being aware of the capabilities of its connection is beneficial in a number of situations.

I am considering how to do this + #72 together since they are somewhat related and thought some discussion would be good first around how the two would interact. Some unordered thoughts about how it would work:

### Defining Capabilities

- `{dialect}+{engine}://...` syntax is nice
  - `postgis+asyncpg://...`
  - `sqlite+aiosqlite://...`
  - `mysql+odbc://...`
  - can still omit the engine to fall back on the default engine for a given dialect: `sqlite://`
- it seems obvious dialects should have a set of supported features defined:
  - postgres `DISTINCT ON` or JSON fields
  - postGIS / spatialite functions
- what about engines? 
  - example, aiosqlite did not support loading sqlite extensions, which is important for spatialite
  - odbc may need to limit capabilities
  - any situations where a client would *add* capabilities?
- if a user picks a silly combination (ex. `postgis+mysql://...`) should we warn them or let it fail? maybe an engine should define which dialects it supports (and postgis would just subclass pg)

### Using Capabilities

- a client is a combination of a dialect and an engine
- capabilities are calculated and kept on the client, representing everything that can be done on the connection
- capabilities can be queried:
  - does this connection support GIS operations?
  - does this connection support table inheritance
- Then, if for example, you were to use a GIS function on a connection that doesn't support it we can be aware of it before making the query.

### Proposed Changes

- rename the backends to the actual engine they are using
  - `mysql` -> `aiomysql`
- move them into an `backends.engines` package
- add a `backends.dialects` package that contains the supported dialects and their features
- create a new client class that "picks" a dialect and an engine
- generate a combined "capabilities" object on the client that passes the capabilities of the dialect through the engine and allow it to remove (or add) things as needed
- allow the query generator to access these capabilities

Then, with the GIS module in mind, it would be nice to be able to register new dialects. For example, having `spatialite` subclass `sqlite` to add the location capability.

**Comment by grigi** [comment]
You are welcome to do it.
I was considering doing this with #72, and have a few thoughts already:

# I'm thinking something along the lines of:
* rename `tortoise.backends` → `tortoise.dialects`
* move the `client.py` out to a `tortoise.connectors` package
* move the `base.config_generator` out to the top package (and extend it to understand dialect/connector)
* create a global `client.py` which is the thing that manages the combination of both the dialect and connector.
* The dialect should have a way to elect a default connector. It should be a function/property, because e.g. we may decide to test if `asyncpg` or `aiopg` is available, and then prefer the one over the other. This could be called from config_generator, as its job is to generate a best effort config. If you are an advanced user, you should know what it is you are doing.
* The dialect should report its supported capabilities
* the connector should report its supported capabilities
* the client reports the intersection of capabilities (with the exception of capabilities that are either client/dialect specific)

Yes to subclassing for `spatialite` from `sqlite`. Same for `postgis`.

# What needs to be a capability:
Some features are done differently, but has the same result. e.g. postgres has a `RETURNING` syntax, and mysql always has a rowid which can be mapped as a serial (autonumber).
We simulate basic JSON persistance for all db's, but we should advertise when we simulate something.

Right now we "simulate":
* `INTERVAL` (everywhere, but we should make it native for Postgres)
* `JSON` (mysql & sqlite)(well actually it is currently still simulated for postgres)
* `DECIMAL` (sqlite)
* `BOOLEAN` (sqlite)

We also "emulate" several comparator functions in mysql.

But generally these emulations work comprehensively, e.g. boolean/decimal. so we need not have caps for those, but others, e.g. indexing/searching inside a JSON which can only be done in one DB should be a capability, but only once we actually implement it.

Other things we want for capabilities is connection settings, e.g. concurrency

And these caps should be made available early, so that even the test runner may skip tests based on them.

ODBC does change some things, such as the parameter syntax. And I think it may have limited support for some datatypes. (e.g. may not work with GIS, but we don't know until we try).

# How to define capabilties:
I'm unsure if it should be a set of caps, or if it should be predicate based?

Some caps are connector specific, such as concurrency. Could even be specific to the configuration of that connector. e.g. concurrency is only supported by versions > X or local config?

SQL Parameter handling changes based on the connector. e.g. `asyncpg` uses `&1` whereas `aiopg` uses `%s` (or `:name`) and `pyodbc` uses `@P1`. So we need to have support for capabilities that is more than just binary, but contains a value.

Based on that I propose caps is a simple dict like so:
```js
{
    "max_concurrency": 1,
    "paramstyle": "ampnum",  // used to lookup a param generator function
    "gis": ["polygon", "point"],  // or could just be true
}
```
And we could just add extra info in there, such as `"db_description": "Postgres v10.2 + Postgis"` etc...

And we don't specify the caps that isn't supported? (or do we? It could clutter...)

# Other:
`aiosqlite` is currently the best asyncio wrapper to sqlite, and we spent a fair bit of time to speed it up. I would prefer that we extend it to support spatialite.

A client could only "add" capabilities that are specific to the client. e.g. concurrency.

Regarding silly combinations... I don't know that we'd buy anything in trying to restrict the user. We could make the connector whitelist its dialects, or we could warn users in the documentation that this is "advanced" settings and we trust they know what they are doing. E.g. when we set up ODBC, we don't really know what is on the other side...

# Test framework
We should be able to skip tests using a simple decorator like so:
```py
@test.skipRequiresCaps(max_concurrency__gt=1, gis=True)
```

**Comment by arlyon** [comment]
Sounds good. Seems like the biggest "challenge" is finding an intuitive (and useful) format for declaring capabilities as well as what actually constitutes one, so that might take a few tries until it feels natural.

As for the emulated functionality, I think it needs more fleshing out. To me, a capability would represent only what that connection handles natively. Tortoise would then, if the capability is not present on the connection, use ea "polyfill" (ie. simulating a JSON field) instead of the native feature via some compat layer.

 I'll take a crack at it this weekend.

**Comment by grigi** [comment]
Fields can be emulated on a per-db level by overriding `to_db_value` and `to_python_value` & overriding the DB field name.
For some emulations such as boolean for SQLite, the emulation is complete, and done entierly in the "dialect" portion. So I feel the dialect need not report that as a missing feature.

I don't want to dynamically polyfill, because each polyfil will likely end up completely unique.

When the emulation cannot be perfect (e.g. JSON) or so complicated we won't even try (e.g. GIS), then we should manage it as a capability.

How we should manage, say, that PosgtreSQL doesn't have `UNSIGNED` integers at all?
Right now we decided against even supporting unsigned integers. But the most correct way would be to use a larger integer field with a check condition, and then WARN about using an unsigned int with PostgreSQL only?
But that can once again provide perfect behaviour, and should we manage that as a capability or just keep it in the dialect?

Hmm, I think I convinced myself. I propose the following rule:
* If we can simulate behaviour perfectly with a polyfill in the dialect only, we don't need a capability fag. (It could however be reported in an informational interface)

**Comment by abondar** [comment]
I am still not quite getting, what such approves gives us?

At the moment it seems like quite big complication for both us, because we will have to support every legit combination of dialect and engine, and for user, who would have to learn what is dialect only to find out that it was introduced to solve problems that he, most likely, will not face. 

I would like to implement more simple approach, where we stick to backends, which store capabilities themselfs.
We also provide possibility to implement custom QuerySet (inherited from base QuerySet) for any connector.
And to solve the problem, that dialects trying to solve, we implement possibility to modify your connector with extension. 

Extension would be pluggable module that can be set one per connection. It's scope of changes would be to implement new `DBClient` class inherited from appropriate connector (e.g. for postgis it would be `AsyncpgDBClient` and to implement custom queryset (which will give ability to change API of QuerySet).

For me it looks like much more straightforward approach, that will cause much less user confusion for users and simpler code for us, giving us approximately same capabilities.  

Or am I missing something and there is problems that such approach can't resolve?

**Comment by arlyon** [comment]
The reasoning on my part for it is that I think dialects and connectors is a better representation of SQL. When I define or extend a dialect, I am specifying or adding new features (and capabilities) to the syntax. When I define an engine / connector, I am specifying which engine to use to execute it. 

Currently our backends are almost like a dialect, and we can define capabilities on them directly, just with a hard-coded connector. If you use postgres, you must use asyncpg, and splitting it out would "fix" that. Though, like you said, this will be unneeded by the vast majority of users, and I may be getting a bit eager.

I'll ignore #72 for now I think, and we can maybe come back to it later if needs be.

**Comment by grigi** [comment]
Closed as per #92 solving this in a minimal way.
